test/clonevolume_tester.py

```diff
@@ -0,0 +1,262 @@
+        mock_3parclient.getTask.return_value = {'status': data.TASK_FAILED}
+
+
+class TestCloneInvalidSourceVolume(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME,
+                         # Difference in size of source and cloned volume
+                         # triggers offline copy. Src volume size is 2.
+                         "size": 20}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        # Source volume not found
+        mock_etcd.get_vol_byname.return_value = None
+
+
+# TODO: Make this fail and in validation compare error message
+class TestCloneWithInvalidSize(CloneVolumeUnitTest):
+    pass
+
+# Online copy with dedup
+class TestCloneDedupVolume(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume_dedup
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+
+
+# Online copy with flash cache flow
+class TestCloneWithFlashCache(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp, {u"Err": ''})
+
+        # Check required WSAPI calls were made
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.createVolumeSet.assert_called()
+        mock_3parclient.modifyVolumeSet.assert_called()
+        mock_3parclient.addVolumeToVolumeSet.assert_called()
+
+    def get_request_params(self):
+        return {"Name": "clone-vol-001",
+                "Opts": {"cloneOf": data.VOLUME_NAME}}
+
+    def setup_mock_objects(self):
+        mock_etcd = self.mock_objects['mock_etcd']
+        mock_etcd.get_vol_byname.return_value = data.volume_flash_cache
+
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.copyVolume.return_value = {'taskid': data.TASK_ID}
+        mock_3parclient.getCPG.return_value = {}
+
+
+# Online copy with flash cache - add to vvset fails
+class TestCloneWithFlashCacheAddVVSetFails(CloneVolumeUnitTest):
+    def check_response(self, resp):
+        self._test_case.assertEqual(resp,
+                                    {u"Err": u'Not found (HTTP 404) - fake'})
+
+        # Check required WSAPI calls were made
+        mock_3parclient = self.mock_objects['mock_3parclient']
+        mock_3parclient.createVolumeSet.assert_called()
```

Comment:
modifyVolumeSet assert call is missing here

Reply:
Added